### ISSUE TITLE
native image build fails - com.oracle.svm.core.util.UserError$UserException: Already registered: com.datastax.oss.driver.shaded.guava.common.primitives.UnsignedBytes$LexicographicalComparatorHolder.getBestComparator()

### ISSUE THREAD

**Issue by thomasbecker** [datastax/cassandra-quarkus#69]
Hey Tomek et al,

thx for all the effort and support to get cassandra-quarkus done! Much appreciated.

I tried to build and use the cassandra-quarkus extension in a new application running on quarkus. Basically it works fine until I try to build a native image which fails with:

```
Error: Already registered: com.datastax.oss.driver.shaded.guava.common.primitives.UnsignedBytes$LexicographicalComparatorHolder.getBestComparator()
com.oracle.svm.core.util.UserError$UserException: Already registered: com.datastax.oss.driver.shaded.guava.common.primitives.UnsignedBytes$LexicographicalComparatorHolder.getBestComparator()
```
It can be fixed by simply removing the corresponding methods in UnsignedBytesSubstitutions.java:

```
 package com.datastax.oss.driver.shaded.guava.common.primitives;

-import com.oracle.svm.core.annotate.Alias;
 import com.oracle.svm.core.annotate.Delete;
-import com.oracle.svm.core.annotate.RecomputeFieldValue;
-import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
-import java.util.Comparator;

 @TargetClass(UnsignedBytes.LexicographicalComparatorHolder.UnsafeComparator.class)
 @Delete
 final class UnsafeComparatorSubstitutions {}

 @TargetClass(UnsignedBytes.LexicographicalComparatorHolder.class)
-final class LexicographicalComparatorHolderSubstitutions {
-
-  @Alias
-  @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.FromAlias)
-  static Comparator<byte[]> BEST_COMPARATOR = UnsignedBytes.lexicographicalComparatorJavaImpl();
-
-  /* All known cases should be covered by the field substitution above... keeping this only
-   * for sake of completeness */
-  @Substitute
-  static Comparator<byte[]> getBestComparator() {
-    return UnsignedBytes.lexicographicalComparatorJavaImpl();
-  }
-}
```

However that just leads to the next problem:
```
Error: Classes that should be initialized at run time got initialized during image building:
 io.smallrye.metrics.interceptors.MetricNameFactory_Bean the class was requested to be initialized at build time (from the command line). io.smallrye.metrics.interceptors.MetricNameFactory_Bean has been initialized without the native-image initialization instrumentation and the stack trace can't be tracked. Try avoiding to initialize the class that caused initialization of io.smallrye.metrics.interceptors.MetricNameFactory_Bean
org.eclipse.microprofile.metrics.MetricRegistry the class was requested to be initialized at build time (from feature io.quarkus.runner.AutoFeature.beforeAnalysis). org.eclipse.microprofile.metrics.MetricRegistry has been initialized without the native-image initialization instrumentation and the stack trace can't be tracked. Try avoiding to initialize the class that caused initialization of org.eclipse.microprofile.metrics.MetricRegistry
```

Need to do some further analysis on the latter issue, but as long as I'm using the metrics dependency:

<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-smallrye-metrics</artifactId>
</dependency>

even if the package is not used in my application someone seems to init both mentioned classes at build time which conflicts with the configuration CassandraClientProcessor.java.

Unfortunately -H:+TraceClassInitialization doesn't seem to work so some detective work will be needed here.

So I guess we need a proper fix for the first issue (or am I doing something wrong here?) and if you have any suggestions on how to find the root cause of the metrics issue that'll be much appreciated.

If there's any way I can support you guys with the extension, I'm happy to do so.

Cheers,
Thomas

**Comment by absurdfarce** [comment]
On the second issue (metrics classes which should be initialized at runtime getting initialized at buildtime): this issue can be reproduced solely by adding smallrye-metrics to the classpath:

```
diff --git a/quickstart/pom.xml b/quickstart/pom.xml
index 3976038..d841c85 100644
--- a/quickstart/pom.xml
+++ b/quickstart/pom.xml
@@ -87,6 +87,10 @@
       <groupId>io.quarkus</groupId>
       <artifactId>quarkus-resteasy-mutiny</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-smallrye-metrics</artifactId>
+    </dependency>
   </dependencies>
   <build>
     <plugins>
```

Metrics do not need to be enabled.

In my case this change results in the following failure for the native-image build:

```
Error: Classes that should be initialized at run time got initialized during image building:
 io.smallrye.metrics.MetricsRegistryImpl the class was requested to be initialized at build time (subtype of org.eclipse.microprofile.metrics.MetricRegistry). To see why io.small
rye.metrics.MetricsRegistryImpl got initialized use -H:+TraceClassInitialization                        
org.eclipse.microprofile.metrics.MetricRegistry the class was requested to be initialized at build time (from feature io.quarkus.runner.AutoFeature.beforeAnalysis). To see why or
g.eclipse.microprofile.metrics.MetricRegistry got initialized use -H:+TraceClassInitialization
```

This is the result of an attempt to build the native image version of the quickstart app using Graal 20.0.0 with local builds of java-driver master and cassandra-quarkus master.

I'm becoming increasingly suspicious of [this bit of code](https://github.com/quarkusio/quarkus/blob/1.5.0.Final/extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java#L143-L158) in the smallrye-metrics extension.  Unless I'm misreading this the extension is specifying a collection of beans which should be added... and it's doing so in the augmentation phase, so this would definitely hit at build time.  Note that MetricNameFactory and MetricRegistries are both present here.  The first is subject to injection (which could easily explain the _Bean extension we see in the error message above) while the second references MetricRegistry directly.

**Comment by absurdfarce** [comment]
Still trying to confirm my analysis, but as a short-term workaround you can comment out the explicit request for runtime initialization of MetricRegistry:

```
diff --git a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraCl
ientProcessor.java
index d7d4618..0357fbe 100644
--- a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
+++ b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
@@ -59,7 +59,6 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
-import org.eclipse.microprofile.metrics.MetricRegistry;
 import org.reactivestreams.Publisher;
 
 class CassandraClientProcessor {
@@ -195,10 +194,12 @@ class CassandraClientProcessor {
     return new RuntimeInitializedClassBuildItem(MetadataManager.class.getCanonicalName());
   }
 
+  /*
   @BuildStep
   RuntimeInitializedClassBuildItem runtimeMetricRegistry() {
     return new RuntimeInitializedClassBuildItem(MetricRegistry.class.getCanonicalName());
   }
+   */
 
   @BuildStep
   NativeImageResourceBuildItem referenceConf() {
```

This got everything working again for both native image and dev mode locally.

Note that you'll probably also have to comment out the other substitution class in UnsignedBytesSubstitution; once I made the change above I started hitting errors due to access of the "deleted" UnsafeComparator class.  Removing the sub which actually marked that as deleted resolved that issue as expected.

**Comment by adutra** [comment]
@absurdfarce @tomekl007 the smallrye-metrics issue is probably a regression introduced by https://github.com/datastax/cassandra-quarkus/pull/61.

**Comment by tomekl007** [comment]
As @absurdfarce correctly pointed out, there is no need to do:
```
@BuildStep
RuntimeInitializedClassBuildItem runtimeMetricRegistry() {
  return new RuntimeInitializedClassBuildItem(MetricRegistry.class.getCanonicalName());
}
```
and introducing it was a mistake.
I have a fix ready:
https://github.com/datastax/cassandra-quarkus/pull/71/files
with new integration tests that catch the problem and will prevent this regression from happening in the future.

**Comment by thomasbecker** [comment]
Quick heads up for you guys. Sorry for the late reply. The end of last week was crazy.

@absurdfarce That's exactly how I fixed it. Removed that configuration from cassandra-quarkus and (for me) it worked. Didn't know if that has any side effects on other scenarios. However it's removed now and that fixes my builld.

I dropped all my changes, rebuild the java-driver and cassandra-quarkus and can confirm that it now works fine when build from the current sources.

Will do further testing and evaluation this week. Let me know if there's anything I can do to help.

Thx for the support and quick fixes, guys! Much appreciated.
